daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -33,30 +33,79 @@ import org.apache.daffodil.util.Maybe
 import org.apache.daffodil.util.Maybe._
 import org.apache.daffodil.util.MaybeULong
 import org.apache.daffodil.util.Misc
+import java.io.File
 
 /**
  * This simple extension just gives us a public method for access to the underlying byte array.
  * That way we don't have to make a copy just to access the bytes.
  */
-private[io] class ByteArrayOutputStreamWithGetBuf() extends java.io.ByteArrayOutputStream {
+private[io] class ByteArrayOrFileOutputStreamWithGetBuf(
+  chunkSizeInBytes: Long = 4096,
+  maxBufferSizeInBytes: Long = 2000 * (1 << 20))
+  extends java.io.ByteArrayOutputStream {
+  private var _byteArrayOS = this
+  private var _fileOS: java.io.FileOutputStream = null
+
+  var isFile: Boolean = false
+  var tempFile: File = null
+  var bytesWritten = count
+
+  def switchOS() = {
+    Assert.invariant(_fileOS == null)
+    tempFile = File.createTempFile("daffodil-", ".tmp")
```

Comment:
There probably needs to be a tunable that says what directory we create this temp file in. Some applications may only be allow to write to certain directories, which might not include /tmp, so this needs to be configurable. Might also be a good idea to make the prefix and suffix tunable as well.

Reply:
I've often wondered if this is tunable at the system level. I.e., can you just setup linux/windows/whatever so that it creates temp files in some location as the default or a per-user setting, like using environment variables or something?

Reply:
Potentially, but I can image use cases were it would be a different directory per-file/job, so a global property might not be sufficient. To accomplish this you could use tunables and change the value before each call to parse.

Reply:
What we might want to do is make is to the default value of the new tunable is the empty string. And we could interpret emptu string as we should use the java.io.tmpdir system property. That way this directly can be controlled globally by setting the system property and setting the tunable to "", or it could be set on a per java, compiler, etc. basis.